### DISCUSSION TITLE
Should ToggleDecorations be CycleDecorations?

### DISCUSSION THREAD

**Discussion by dashavoo** [labwc/labwc#2070] [General]
Creating this as a discussion point rather than a bug as the documentation is clear, it just took me a long time to think of looking there.
A toggle is usually a boolean switch between two states, but in labwc it is three-states. ToggleDecorations is also not backwards compatible with the Openbox action of the same name, which is two-way only.
Would CycleDecorations be a better name for the three-state-switch, with ToggleDecorations behaving like Openbox?
If not (because probably backwards compatibility between labwc versions is more important than openbox compatibility), what would be an appropriate name for an Action that behaves like the original Openbox version?

**Comment by johanmalm** [comment]
Open to ideas. Ideally we shouldn’t break backward compatibility, but we have to weight up against cleanliness if we can see a neater approach.

**Comment by jp7677** [comment]
Somewhat related: As a follow up to https://github.com/labwc/labwc/issues/1985 I thought about changing the `ToggleTabletMouseEmulation` so that it can be used as a window rule. Therefore I can:
- Extend `ToggleTabletMouseEmulation` with a `state` like parameter for `on|off|toggle`. Should it be `CycleTabletMouseEmulation` in that case?
- Introduce a new action `SetTabletMouseEmulation` with a `state` like parameter for `on|off`
- Introduce two new actions for `EnableTabletMouseEmulation` and `DisableTabletMouseEmulation`

I think we have examples of all version, which one would be the most consistent with the existing options?

**Comment by droc12345** [comment]
I think the name is fine, as it's like openbox.
OB cycles between on and off with off being modified by the keepborder setting, ie if set you will have either a border or border and titlebar.
So the labwc version just enhances the OB version, by allowing on/border/off 
Even if we implement Decorate/Undecorate actions, they still use keepborder to make decisions, at least that's how the OB docs read to me.l

**Comment by dashavoo** [comment]
It doesn't enhance it. In Openbox, regardless of whether you have keepborder, there are only two states. This is consistent with the name "toggle". The user can choose whether or not the border is kept when removing decorations.

It might be easier to explain why it is inconsistent with an example:
I have a key binding which toggles the active window being maximised and toggles the decorations. This behaves differently to making a window fullscreen. In openbox I could use this binding with keepborder. In labwc I can't. Behaviour without keepborder is fine, but with it turned on the state of the window gets out of sync and I quickly have either a maximised window with a titlebar, or a non-maximised one without.

**Reply by droc12345** [reply]
That doesn't sound like it's the toggling that's the problem. Sounds like it might be in the maximize window area.

Yes in open box you have 2 states
keepborder on - toggle is border or border with titlebar
keepborder off - toggle is undecorated or border with titlebar

Labwc
keepborder on - 3 states, full decorations, border only, undecorated
keepborder off - 2 states full decorations or undecorated.

The reason I say enhancement, is that I don't have to do a reconfigure to modify windows decorations.

Maximize leaves the titlebar alone, but the border seems to disappear, which I don't think is correct. 

But in any case, I can't get things to get out of sync on my end.

**Reply by dashavoo** [reply]
> But in any case, I can't get things to get out of sync on my end.

I guess I haven't explained well enough then.

Add a keybinding such as this to your config:
```xml
    <keybind key="F11">
      <action name="ToggleMaximize"/>
      <action name="ToggleDecorations"/>
    </keybind>
```

The table below shows behaviour in openbox and labwc - hopefully you can see here what I mean about the labwc 3-state switch getting out of sync.

  | openbox (no keepBorder) | labwc (no keepborder) | openbox (yes keepBorder) | labwc (yes keepBorder)
-- | -- | -- | -- | --
Initial state | unmaximised<br> title<br> border | unmaximised<br> title<br> border | unmaximised<br> title<br> border | unmaximised<br> title<br> border
Press 1 | :heavy_check_mark:<br>maximised<br> no title<br> no border | :heavy_check_mark:<br>maximised<br> no title<br> no border | :heavy_check_mark:<br>maximised<br> no title<br> border | :heavy_check_mark:<br>maximised<br> no title<br> border
Press 2 | :heavy_check_mark:<br>unmaximised<br> title<br> border     | :heavy_check_mark:<br>unmaximised<br> title<br> border     | :heavy_check_mark:<br>unmaximised<br> title<br> border  | :x:<br>unmaximised<br> no title<br> no border
Press 3 | :heavy_check_mark:<br>maximised<br> no title<br> no border | :heavy_check_mark:<br>maximised<br> no title<br> no border | :heavy_check_mark:<br>maximised<br> no title<br> border | :x:<br>maximised<br> title<br> border
Press 4 | :heavy_check_mark:<br>unmaximised<br> title<br> border     | :heavy_check_mark:<br>unmaximised<br> title<br> border     | :heavy_check_mark:<br>unmaximised<br> title<br> border  | :x:<br>unmaximised<br> no title<br> border
Press 5 | :heavy_check_mark:<br>maximised<br> no title<br> no border | :heavy_check_mark:<br>maximised<br> no title<br> no border | :heavy_check_mark:<br>maximised<br> no title<br> border | :x:<br>maximised<br> no title<br> no border


> The reason I say enhancement, is that I don't have to do a reconfigure to modify windows decorations.

This could be achieved with a separate `ToggleBorder` action. You could also have a separate `ToggleTitleBar` action.

**Comment by droc12345** [comment]
I see what the problem is, it's not that things are out of sync, as far as labwc is concerned, it's
because of the double action binding, with one being 3 states and the other 2. 
The end result is not surprising.

Edit to add: OB has a couple of actions we haven't implemented yet,  Decorate and Undecorate which shouldn't be too hard
to implement, and I have considered creating a 3rd action, for just the border, though that one might be a little more tricky.

**Comment by droc12345** [comment]
I have a question (might be a silly one) why are you not using fullscreen instead of maximize as that automatically removes decorations?

**Reply by dashavoo** [reply]
There are differences between the specification for fullscreen and maximised.

One that applies in openbox but doesn't seem to in in labwc (don't know whether by design or not) is that fullscreen applications appear above "always on top" ones.

Some differences are application dependent, for example Firefox fullscreen hides the address bar, etc., whereas maximised keeps them.

**Reply by droc12345** [reply]
Not sure about the design, but IMO, fullscreen should be above anything else, otherwise other windows would be in the way.

The only visible differences that I see between max and full is ...
Full removes any opacity, full doesn't show a titlebar or borders and full covers all the screen including exclusive zones, ie panels. 

Anyway, I'm not sure they will fix the 3 state, so those who depend on 2 state behavior will have to find some way around the "enhancement"

**Comment by droc12345** [comment]
Have added a PR for adding actions, Decorate, Undecorate and Border. #2124

Edit to add: this does not undo the 3 way switch, as I'm not sure whether they really want to go that way.
I just added extra options.

**Reply by dashavoo** [reply]
Yeah, I opened a discussion rather than an issue because even though removing the three-way switch would fix backwards compatibility with openbox it would be a breaking change for labwc. I think it would be better to go back to openbox compatibility, but its not my call.

Thanks for adding more possibilities. I'll have a look at the PR.

**Reply by droc12345** [reply]
Never mind, they had something similar though it doesn't follow the openbox way, so I closed that PR.

It's a little hard to keep track of when to follow openbox and when to ignore it.

**Reply by droc12345** [reply]
If you can apply patches and you use labwc git, I have started to carry patches on my own
https://github.com/droc12345/images/blob/main/patches/tog-decor-fix.patch 
-- changes 3 way togggleDecor to 2 way like openbox

**Comment by droc12345** [comment]
Maybe a poll to find out how many actually use the 3 way switch.

Other than playing around to figure out what the problem was, 
I typically run with just borders, set with a "*" window rule so all windows look the same.
And I have a feeling that most are that way, they decide what ssd stuff they want and hardly ever change.
